### DISCUSSION TITLE
Where to download binaries?

### DISCUSSION THREAD

**Discussion by mwitteveen** [prisma/prisma#14873] [Q&A] [answered]
I need to download the migration binary for a different system that I am on. Where do I do this? 

- I tried to build them from scratch. 
- spoofing my architecture when I'm installing

I do not have internet access on the target machine so I can't build them on there. 

My dev machine is darwin Mac os m1, the VM openssl 1.0 ubuntu, my linux PC openssl 1.1 ubuntu and the CICD pipeline rhel with openssl 3.0. I need to download them externally.

**Comment by austincrim** [comment, marked as answer]
Hey @mwitteveen 👋,

Have you tried specifying multiple [`binaryTargets`](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#binarytargets-options) in your Prisma schema's `generator` block? After running `npx prisma generate`, you should be able to find the engines in `node_modules/.prisma/client`.

**Reply by vladi-stevanovic** [reply]
👋 Hello @mwitteveen I'm going through pending GitHub Discussions to check if you need any help. Is this issue still open?
Please let us know if we can help you further or if Austin's suggestion help, feel free to "Mark as answer". 

It would be super useful for other users facing this problem! 🙌

**Reply by mwitteveen** [reply]
Hi @austincrim yes but this is not working when I have no access to internet on the machine where I am generating the client. So I am copying my code and than developing in a no internet environment. Thus these generators will not be downloaded.

**Reply by mwitteveen** [reply]
@vladi-stevanovic no so in the end i am creating a RHEL docker container on my linux laptop ( that has the same architecture) then ssh into that container. Do a npm install @prisma/engines globally and going to the node_modules/@prisma/engines folder. This now contains the binaries. Then I copy them onto my host machine, email them to myself. Get the binaries on the machine where I am running the VM on. Then copy them onto the VM. Then on the VM changing the mode to 777 (with chmod 777 -R <binaries folder>) and then setting the environment variables in my CICD pipeline to the binaries that I copied.

**Reply by mwitteveen** [reply]
-.-'

**Reply by mwitteveen** [reply]
@vladi-stevanovic so i am stuck at one version for a while

**Reply by pantharshit00** [reply]
You can download engines by using the following URL:
`https://binaries.prisma.sh/all_commits/{COMMIT_ID}/{PLATFORM}/{ENGINE_NAME}.gz`

where
**COMMIT_ID:** Full commit hash from https://github.com/prisma/prisma-engines. You can check the tags page for specific commit which relates to the release you want: https://github.com/prisma/prisma-engines/tags
**PLATFORM:** Platform of your choice. Full list here: https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#binarytargets-options 
**ENGINE_NAME**: Engine you want to download. One of `libquery_engine.so.node`, `query-engine`, `migration-engine`, `introspection-engine` and `prisma-fmt`. 

Example:
https://binaries.prisma.sh/all_commits/694eea289a8462c80264df36757e4fdc129b1b32/debian-openssl-3.0.x/libquery_engine.so.node.gz


Let me know if this helps.

**Reply by pantharshit00** [reply]
You can also write a small script which can fetch these for you where you can use the GitHub API to query the correct commit_id.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
